File: src/simulation/space/resource/dependency/NeedDependency.java
package simulation.space.resource.dependency;

import simulation.space.resource.tag.labeler.ResourceLabeler;
import simulation.space.tile.Tile;
import simulation.space.resource.Resource;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class NeedDependency extends LabelerDependency {
    public Set<String> lastConsumed = new HashSet<>();

    public NeedDependency(
            double amount,
            double deprivationCoefficient,
            boolean isNecessary,
            ResourceLabeler goodResources
    ) {
        super(deprivationCoefficient, isNecessary, amount, goodResources);
    }

    @Override
    double satisfaction(Tile tile, Resource resource) {
        double _amount = amount * (resource == null ? 1 : resource.getAmount());
        int currentAmount = 0;
        for (Resource res : tile.getAccessibleResources()) {
            if (res.equals(resource)) {
                continue;
            }
            if (isResourceDependency(res)) {
                currentAmount += res.getAmount();
                if (currentAmount >= _amount) {
                    break;
                }
            }
        }
        return Math.min(((double) currentAmount) / _amount, 1);
    }

    @Override
    public boolean hasNeeded(Tile tile) {
        return tile.getAccessibleResources().stream().anyMatch(this::isResourceDependency);
    }

    @Override
    public boolean isPositive() {
        return true;
    }
}

File: src/simulation/space/resource/dependency/NeedDependency.kt
package simulation.space.resource.dependency

import simulation.space.resource.Resource
import simulation.space.resource.tag.labeler.ResourceLabeler
import simulation.space.tile.Tile
import kotlin.math.min


class NeedDependency(
        amount: Double,
        deprivationCoefficient: Double,
        isNecessary: Boolean,
        goodResources: ResourceLabeler
) : LabelerDependency(deprivationCoefficient, isNecessary, amount, goodResources) {
    var lastConsumed = mutableSetOf<String>()

    override fun satisfaction(tile: Tile, resource: Resource): Double {
        val actualAmount = amount * resource.amount
        var currentAmount = 0

        loop@for (list in tile.accessibleResources)
            for (res in list) {
                if (res == resource)
                    continue

                if (isResourceDependency(res)) {
                    currentAmount += res.amount * oneResourceWorth(res)
                    if (currentAmount >= actualAmount)
                        break@loop
                }
            }

        return min(currentAmount.toDouble() / actualAmount, 1.0)
    }

    override fun hasNeeded(tile: Tile) =
            tile.accessibleResources.any { it.asSequence().any { r -> isResourceDependency(r) } }

    override val isPositive = true
}
